src/main/java/tech/pegasys/tools/epchecks/JavaCase.java

```diff
@@ -0,0 +1,160 @@
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.ModifiersTree;
+import com.sun.source.tree.VariableTree;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+    name = "JavaCase",
+    summary = "Use the appropriate case.",
+    severity = SUGGESTION,
+    linkType = BugPattern.LinkType.NONE)
+public class JavaCase extends BugChecker
+    implements MethodTreeMatcher, ClassTreeMatcher, VariableTreeMatcher {
+
+  private static final Pattern PATTERN_HAS_LOWER = Pattern.compile("^.*[a-z].*$");
+  private static final Pattern PATTERN_HAS_UPPER = Pattern.compile("^.*[A-Z].*$");
+  private static final Pattern PATTERN_STARTS_WITH_LOWER = Pattern.compile("^[a-z].*$");
+  private static final Pattern PATTERN_STARTS_WITH_UPPER = Pattern.compile("^[A-Z].*$");
+  private static final Pattern PATTERN_UNDERSCORES = Pattern.compile("^_+$");
+
+  @Override
+  public Description matchVariable(VariableTree tree, VisitorState state) {
+    String name = tree.getName().toString();
+    ModifiersTree modifiers = tree.getModifiers();
+    Set<Modifier> flags = modifiers.getFlags();
+
+    // Constants should be all uppercase.
+    if (flags.contains(Modifier.STATIC) && flags.contains(Modifier.FINAL)) {
+      if (!isUpperUnderscore(name)) {
+        return buildDescription(tree)
+            .addFix(SuggestedFixes.renameVariable(tree, toUpperUnderscore(name), state))
+            .build();
+      }
+    } else if (!isUnderscore(name) && !isLowerCamel(name)) {
+      return buildDescription(tree)
+          .addFix(SuggestedFixes.renameVariable(tree, toLowerCamel(name), state))
+          .build();
+    }
+    return Description.NO_MATCH;
+  }
+
+  @Override
+  public Description matchMethod(MethodTree tree, VisitorState state) {
+    String name = tree.getName().toString();
+
+    // Class constructors have the <init> name.
+    if (name.equals("<init>")) return Description.NO_MATCH;
+
+    // Test methods often have underscores which are fine.
+    if (!isLowerCamel(name) && !isTestMethod(tree)) {
+      return buildDescription(tree)
+          .addFix(SuggestedFixes.renameMethod(tree, toLowerCamel(name), state))
+          .build();
+    }
+    return Description.NO_MATCH;
+  }
+
+  @Override
+  public Description matchClass(ClassTree tree, VisitorState state) {
+    String name = tree.getSimpleName().toString();
+
+    // Anonymous classes have no name.
+    if (name.isEmpty()) {
+      return Description.NO_MATCH;
+    }
+
+    if (!isAllUpper(name) && !isUpperCamel(name)) {
+      String update = tree.toString().replace(name, toUpperCamel(name));
+      update = update.split("\\r?\\n")[1];
+      return buildDescription(tree)
+          .addFix(SuggestedFix.builder().replace(tree, update).build())
+          .build();
+    }
+    return Description.NO_MATCH;
+  }
+
+  private static boolean isUnderscore(String name) {
+    return PATTERN_UNDERSCORES.matcher(name).matches();
+  }
```

Comment:
I'm not sure we need to use a regex for this
```suggestion
  private static boolean isUnderscore(String name) {
    return "_".equals(name);
  }
```

would be faster and more readable.

Reply:
Ah that regex actually matches one or more underscores. There are often `__` variables in lambdas. I will rename the function to be more clear.